File: data_structures/stacks_and_queues/stack.py
class Node():
    def __init__(self, value):
        self.value = value
        self.next = None

# Stack - Array implementation
class Stack():
    def __init__(self):
        self.array = []
        
    def __str__(self):
        return str(self.__dict__)
        
    def peek(self):
        return self.array[len(self.array)-1]
        
    def push(self, value):
        self.array.append(value)
        return self
        
    def pop(self):
        self.array.pop()
        return self

# Stack - Linked List implementation
class Stack():
    def __init__(self):
        self.top = None
        self.bottom = None
        self.length = 0
        
    def __str__(self):
        return str(self.__dict__)
    
    def peek(self):
        return self.top.value
        
    def push(self, value):
        new_node = Node(value)
        if self.top == None:
            self.bottom = new_node
            self.top = self.bottom
            self.length = 1
        else:
            holding = self.top
            self.top = new_node
            self.top.next = holding
            self.length += 1
            
    def pop(self):
        if not self.top:
            return None
        if self.top == self.bottom:
            self.bottom = None
        self.top = self.top.next
        self.length -= 1
        return self
    
    def printlist(self):
        myarr = []
        currentNode = self.top
        while currentNode != None:
            myarr.append(currentNode.value)
            currentNode = currentNode.next
        return myarr
        
stack = Stack()
stack.push(4)
stack.push(3)
stack.push(2)
stack.push(1)
stack.pop()
stack.pop()
print(stack.printlist())
print(stack.peek())

File: data_structures/stacks_and_queues/Stack.java
package data_structures.stacks_and_queues;

class Node {
    String value;
    Node next;
    Node(String value) {
        this.value = value;
        this.next = null;
    }
}

class Stack {
    private Node top;
    private Node bottom;
    private int length;

    Stack() {
        this.top = null;
        this.bottom = null;
        this.length = 0;
    }

    public String peek() {
        return top != null ? top.value : null;
    }

    public void push(String value) {
        Node newNode = new Node(value);
        if (length == 0) {
            top = newNode;
            bottom = newNode;
        } else {
            Node holding = top;
            top = newNode;
            top.next = holding;
        }
        length++;
    }

    public void pop() {
        if (top == null) {
            return;
        }
        if (top == bottom) {
            bottom = null;
        }
        top = top.next;
        length--;
    }

    public String[] printList() {
        String[] result = new String[length];
        int index = 0;
        Node currentNode = top;
        while (currentNode != null) {
            result[index++] = currentNode.value;
            currentNode = currentNode.next;
        }
        return result;
    }

    // Returns a string representation of the queue for debugging purposes
    @Override
    public String toString() {
        StringBuilder result = new StringBuilder("{");
        result.append("\"top\": ").append(top != null ? top.value : "null").append(", ");
        result.append("\"bottom\": ").append(bottom != null ? bottom.value : "null").append(", ");
        result.append("\"length\": ").append(length);
        result.append("}");
        return result.toString();
    }

    public static void main(String[] args) {
        Stack stack = new Stack();
        stack.push("4");
        System.out.println(java.util.Arrays.toString(stack.printList()));
        stack.push("3");
        System.out.println(java.util.Arrays.toString(stack.printList()));
        stack.push("2");
        System.out.println(java.util.Arrays.toString(stack.printList()));
        stack.push("1");
        System.out.println(java.util.Arrays.toString(stack.printList()));
        stack.pop();
        System.out.println(java.util.Arrays.toString(stack.printList()));
        System.out.println("Peek: " + stack.peek());
        System.out.println("toString: " + stack.toString());
    }
}
